### ISSUE TITLE
Warn customer if they are on an older version of core-tools and ask them to upgrade

### ISSUE THREAD

**Issue by kulkarnisonia16** [Azure/azure-functions-core-tools#1855]
We can define what older means (it can be based on time like 3-6 months older, # versions older or last known version which we know would still work). We might be able to avoid support cases that are opened because of this - example - https://servicedesk.microsoft.com/#/customer/cases?caseNumber=120012422002597

**Comment by anthonychu** [comment]
Use feed to determine latest version. https://functionscdn.azureedge.net/public/cli-feed-v3.json

Need to discuss what the message looks like and how frequently to show it (probably not on every invocation).

**Comment by anthonychu** [comment]
Here's a starting point for discussion.

* On `func start` and `func help`, display a warning message if there's a newer release of the current major version (i.e., for v2, only check if there's a newer v2)
* Use the feed to determine latest versions https://functionscdn.azureedge.net/public/cli-feed-v3.json
* Add support for an environment variable that can disable the warning (VS and VS Code can set this as they manage their own version or does their own version checks, a customer can use this to silence the check as well)

Would be great if the message can be displayed near the bottom of the output. For `help`, display it at the end. For `start`, display as close to the `Host lock lease acquired by instance ID` message as possible (maybe after the ASP.NET generated `Application started. Press Ctrl+C to shut down.` output).

Checking every time could be costly (also consider if there's no internet connection). If there's a way to check only once per day, that'd be great. But if we don't currently have a way of saving this sort of info, it should be okay checking on every run as long as we can guarantee it's a quick call and handles failures/timeouts silently.

We can use a similar message to [VS Code](https://github.com/microsoft/vscode-azurefunctions/blob/9f5995d429a2aed9f8f7fa1772db3f5c94767036/src/funcCoreTools/validateFuncCoreToolsIsLatest.ts#L79):

> Update your Azure Functions Core Tools ({0}) to the latest ({1}) for the best experience. (followed by an aka.ms link or upgrade instructions if we can figure out what to display for the platform/installer)

@kashimiz @ankitkumarr @fabiocav @ahmelsayed @soninaren Would love your thoughts.

+ @EricJizbaMSFT for whether this affects VS Code and any learnings from implementing this in VS Code.

**Comment by soninaren** [comment]
This sounds good. All the tools i know use `cli-feed-v3.json` to determine the latest version of core tools, so i don't think there will be conflict. Have one question. Where would you show the warning in `func start`? With all the output in the `func start` a warning would just disappear very quickly.

 A possible addition would be a self update gesture at the end of the warning. However implementing that would be a much bigger task.

**Comment by kashimiz** [comment]
@anthonychu Do we have to check `cli-feed-v3.json` every time? We could retrieve the latest value at startup and cache it in  memory, as a start.

@soninaren We had also discussed showing the warning when other commands are executed, ex. every invocation. Definitely more noticeable, but possibly an over-correction. What update reminder frequency have you found effective for command-line tools?

**Comment by soninaren** [comment]
@kashimiz, I don't think we have update reminder any where other than VS Code. They might have telemetry data for update reminder that we could use.

**Comment by anthonychu** [comment]
@kashimiz Not sure where we could stash this info. In another issue @soninaren and I were wondering if we could download bundles to ~/.azure-functions-core-Tools. Perhaps if we standardize on this we can stash a file in there as well.

@soninaren we can display it as late as possible at startup. If it’s a warning, it should show in a another color. Hopefully when #1131 is completed it’ll be even more noticeable.

**Comment by EricJizbaMSFT** [comment]
Actually don't collect much telemetry on this in VS Code. Could add some, but seems like you might implement this feature before we got any meaningful data anyways. Here's a summary of what we do:

We check every time our extension is activated in VS Code (which happens automatically if they have a "host.json" file in their workspace or if they interact with our Functions view). We only prompt if they already have "func" installed and the feed shows a newer version matching their major version (we don't prompt to upgrade across major versions). Finally, we include a "Don't warn again" option in case users want to ignore this.

On another note, if you settle on a standard folder, can it be ".func" or something shorter? ".azure-functions-core-tools" would be the longest folder in my home directory by a decent margin...

**Comment by soninaren** [comment]
> @soninaren we can display it as late as possible at startup. If it’s a warning, it should show in a another color. Hopefully when #1131 is completed it’ll be even more noticeable.

@anthonychu This would be very difficult to time, since the logs in start path come from a lot of different sources. Also if the customer has pending executions, the warning may not even be visible for more than a 1 second.  The executions would push this line up in the feed very quickly. 

One way to get around this would be have a prompt instead that says something like `Using old version of Core-tools press any to continue or run npm ** to install the latest version of coretools`. However this might break automation scenarios for the customers.

**Comment by anthonychu** [comment]
Press key to continue is probably too intrusive. Perhaps we can display it at the beginning of every command, pausing for about a second before proceeding. Maybe that's enough to catch their attention.

**Comment by soninaren** [comment]
> Press key to continue is probably too intrusive. Perhaps we can display it at the beginning of every command, pausing for about a second before proceeding. Maybe that's enough to catch their attention.

I think that would work.

**Comment by fabiocav** [comment]
Need to follow up with @anthonychu on priority compared to https://github.com/Azure/azure-functions-core-tools/issues/2091

**Comment by anthonychu** [comment]
#2091 has higher priority.

**Comment by khkh-ms** [comment]
PR: https://github.com/Azure/azure-functions-core-tools/pull/3327

**Comment by mattchenderson** [comment]
This item was resolved by that PR
